core/src/main/java/io/questdb/griffin/DatabaseSnapshotAgent.java

```diff
@@ -0,0 +1,297 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin;
+
+import io.questdb.cairo.*;
+import io.questdb.cairo.sql.Record;
+import io.questdb.cairo.sql.RecordCursor;
+import io.questdb.cairo.vm.Vm;
+import io.questdb.cairo.vm.api.MemoryCMARW;
+import io.questdb.griffin.engine.table.TableListRecordCursorFactory;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.*;
+import io.questdb.std.str.Path;
+import org.jetbrains.annotations.TestOnly;
+
+import java.io.Closeable;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class DatabaseSnapshotAgent implements Closeable {
+
+    private final static Log LOG = LogFactory.getLog(DatabaseSnapshotAgent.class);
+
+    private final CairoEngine engine;
+    private final CairoConfiguration configuration;
+    private final FilesFacade ff;
+    private final Path path = new Path();
+    private boolean snapshotInProgress;
+    // List of readers kept around to lock partitions while a database snapshot is being made.
+    private final ObjList<TableReader> snapshotReaders = new ObjList<>();
+
+    public DatabaseSnapshotAgent(CairoEngine engine) {
+        this.engine = engine;
+        this.configuration = engine.getConfiguration();
+        this.ff = configuration.getFilesFacade();
+    }
+
+    @Override
+    public synchronized void close() {
+        Misc.free(path);
+        releaseReaders();
+    }
+
+    @TestOnly
+    public synchronized void clear() {
+        snapshotInProgress = false;
+        releaseReaders();
+    }
+
+    @TestOnly
+    public synchronized void releaseReaders() {
+        Misc.freeObjList(snapshotReaders);
+        snapshotReaders.clear();
+    }
+
+    public synchronized void prepareSnapshot(SqlExecutionContext executionContext) throws SqlException {
```

Comment:
This is not a lightweight method, can take a while to execute. It would be better instead of making it `synchronised` use other ways to make it thread safe, for example using some sort of CAS lock to throw `already in progress` sooner instead of waiting.

Reply:
I don't think someone will be calling SNAPSHOT PREPARE/COMPLETE concurrently and, even if they do, they should be ok with a few additional seconds required to learn that there is concurrent snapshot statement execution. Nevertheless, I've replaced synchronized methods with `ReentrantLock` + its optimistic locking API.